blazar/plugins/oshosts/host_plugin.py

```diff
@@ -1005,21 +1006,31 @@ def _poll_resource_failures(self):
         :return: a list of failed hosts, a list of recovered hosts.
         """
         hosts = db_api.host_get_all_by_filters({})
-        reservable_hosts = [h for h in hosts if h['reservable'] is True]
-        unreservable_hosts = [h for h in hosts if h['reservable'] is False]
-
+        
+        ironic_hosts = []
+        nova_hosts = []
+        for h in hosts:
+            if h['hypervisor_type'] == 'ironic': ironic_hosts.append(h)
+            else: nova_hosts.append(h)
+
+        failed_hosts = []
+        recovered_hosts = []
         try:
-            hvs = self.nova.hypervisors.list()
+            if ironic_hosts:
+                ironic_client = ironic.BlazarIronicClient()
+                nodes = ironic_client.ironic.node.list()
+                failed_nodes = [n.uuid for n in nodes 
+                                if n.maintenance or n.power_state in ['error'] 
+                                or n.provision_state not in ['active', 'available']]
+                failed_hosts.extend([h for h in ironic_hosts if h['hypervisor_hostname'] in failed_nodes])
+
+            if nova_hosts:
+                hvs = self.nova.hypervisors.list()
 
-            failed_hv_ids = [str(hv.id) for hv in hvs
-                             if hv.state == 'down' or hv.status == 'disabled']
-            failed_hosts = [host for host in reservable_hosts
-                            if host['id'] in failed_hv_ids]
+                failed_hv_ids = [str(hv.id) for hv in hvs
+                                 if hv.state == 'down' or hv.status == 'disabled']
+                failed_hosts.extend([h for h in nova_hosts if h['id'] in failed_hv_ids])
 
-            active_hv_ids = [str(hv.id) for hv in hvs
-                             if hv.state == 'up' and hv.status == 'enabled']
-            recovered_hosts = [host for host in unreservable_hosts
```

Comment:
Hrm, I see there was prior code for treating unreservable hosts as recovered. I wonder why that is.

Reply:
It checks if those unreservable hosts become "normal". If so, it will automatically put them back to the reservation pool.

Reply:
Ahh, I understand now. The `reservable` property is in fact _how_ it knows if resources are valid in a lease as well. In that case... I was wrong to suggest that we remove this code. Can you please restore it, and the equivalent for the Ironic case? Sorry about that.